proxy/src/coco.rs

```diff
@@ -0,0 +1,467 @@
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Branch(pub String);
+
+/// Tag name representation.
+///
+/// We still need full tag support.
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Tag(pub String);
+
+/// Representation of a person (e.g. committer, author, signer) from a repository. Usually
+/// extracted from a signature.
+pub struct Person {
+    /// Name part of the commit signature commit.
+    pub name: String,
+    /// Email part of the commit signature commit.
+    pub email: String,
+    /// Reference (url/uri) to a persons avatar image.
+    pub avatar: String,
+}
+
+/// Representation of a code commit.
+pub struct Commit {
+    /// Identifier of the commit in the form of a sha1 hash. Often referred to as oid or object
+    /// id.
+    pub sha1: String,
+    /// The author of the commit.
+    pub author: Person,
+    /// The summary of the commit message body.
+    pub summary: String,
+    /// The entire commit message body.
+    pub message: String,
+    /// The recorded time of the committer signature. This is a convenience alias until we
+    /// expose the actual author and commiter signatures.
+    pub committer_time: git2::Time,
+}
+
+impl From<&surf::git::Commit> for Commit {
+    fn from(commit: &surf::git::Commit) -> Self {
+        let mut s = DefaultHasher::new();
+        commit.author.email.hash(&mut s);
+
+        let avatar = format!(
+            "https://avatars.dicebear.com/v2/jdenticon/{}.svg",
+            s.finish().to_string()
+        );
+
+        Self {
+            sha1: commit.id.to_string(),
+            author: Person {
+                name: commit.author.name.clone(),
+                email: commit.author.email.clone(),
+                avatar,
+            },
+            summary: commit.summary.clone(),
+            message: commit.message.clone(),
+            committer_time: commit.author.time,
+        }
+    }
+}
+
+/// Git object types.
+///
+/// `shafiul.github.io/gitbook/1_the_git_object_model.html`
+#[derive(Debug, Eq, Ord, PartialOrd, PartialEq)]
+pub enum ObjectType {
+    /// References a list of other trees and blobs.
+    Tree,
+    /// Used to store file data.
+    Blob,
+}
+
+/// Set of extra information we carry for blob and tree objects returned from the API.
+pub struct Info {
+    /// Name part of an object.
+    pub name: String,
+    /// The type of the object.
+    pub object_type: ObjectType,
+    /// The last commmit that touched this object.
+    pub last_commit: Option<Commit>,
+}
+
+/// File data abstraction.
+pub struct Blob {
+    /// Best-effort guess if the content is binary.
+    pub binary: bool,
+    /// Actual content of the file, if the content is ASCII.
+    pub content: Option<String>,
+    /// Extra info for the file.
+    pub info: Info,
+}
+
+/// Result of a directory listing, carries other trees and blobs.
+pub struct Tree {
+    /// Absolute path to the tree object from the repo root.
+    pub path: String,
```

Comment:
I wonder why we use `String` instead of `PathBuf` here and in a few other places, since PathBuf should more accurately represent the data, as well as represent all paths, which String isn't guaranteed to.

Reply:
Leftovers from when these structs where used directly for the GraphQL API.